### PR TITLE
Implement state file output for RGBASM

### LINKED ISSUE #606: Add constants to sym file as well
What makes the SYM file useful is that it's not easy to know where a given label is located; however, not all constants are obvious, as in the following example.
```asm
SECTION "Gfx", ROMX
Gfx:
INCBIN "gfx.2bpp"

GFX_SIZE equ @ - Gfx ; What value do I have?
    PRINTT "{GFX_SIZE}\n" ; This can help, but it's easy to lose it in a busy `make` invocation
```
For compatibility with existing implementations, maybe output them in comments, at the end of the file, with a special comments preceding them?

Note that this ties to #483.

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #606

```console
% cat foo.asm
DEF x EQU 1
def y = 2
Def w Equs "hello"
charmap "z", 4
MACRO foo
ENDM
Def p Equ 1000
def q = 2000
DEF r EQUS "a\tb\nc\0!"
charmap "{r}", 3000, 4000, 5000, 6000, 1234567890
macro bar
  dw \#
endm
% ./rgbasm -s all:- foo.asm
; File generated by rgbasm

; Numeric constants
def x equ $1
def p equ $3e8

; Variables
def y = $2
def q = $7d0

; String constants
def w equs "hello"
def r equs "a\tb\nc\0!"

; Character maps
newcharmap main
charmap "z", $4
charmap "a\tb\nc\0!", $bb8, $fa0, $1388, $1770, $499602d2

; Macros
macro foo
endm
macro bar
  dw \#
endm
```

### CODE REVIEW

File: include/asm/symbol.hpp [hunk 1]
```diff
@@ -44,7 +44,8 @@ struct Symbol {
 	    >
 	    data;
 
-	uint32_t ID; // ID of the symbol in the object file (-1 if none)
+	uint32_t ID;  // ID of the symbol in the object file (-1 if none)
+	uint32_t key; // Ordering of the symbol in the state file
```
Comment (Rangi42) [inline include/asm/symbol.hpp:48]:
```suggestion
	uint32_t defIndex; // Ordering of the symbol in the state file
```

File: include/asm/output.hpp [hunk 2]
```diff
@@ -21,5 +31,6 @@ void out_CreateAssert(
     AssertionType type, Expression const &expr, std::string const &message, uint32_t ofs
 );
 void out_WriteObject();
+void out_WriteState(std::string name, std::vector<StateFeature> const &features);
```
Comment (Rangi42) [inline include/asm/output.hpp:34]: Why is `name` passed by value?

File: man/rgbasm.1 [hunk 3]
```diff
@@ -86,6 +87,38 @@ Write
 .Xr make 1
 dependencies to
 .Ar depend_file .
+.It Fl s Ar features Ns Op : Ns Ar state_file , Fl \-state Ar features Ns Op : Ns Ar state_file
```
Comment (Rangi42) [inline man/rgbasm.1:90]: This isn't in alphabetical order!

Comment (Rangi42) [inline man/rgbasm.1:90]:
```suggestion
.It Fl s Ar features Ns : Ns Ar state_file , Fl \-state Ar features Ns : Ns Ar state_file
```

File: src/asm/main.cpp [hunk 4]
```diff
@@ -89,14 +91,15 @@ static void printUsage() {
 	    "Usage: rgbasm [-EVvw] [-b chars] [-D name[=value]] [-g chars] [-I path]\n"
 	    "              [-M depend_file] [-MG] [-MP] [-MT target_file] [-MQ target_file]\n"
 	    "              [-o out_file] [-P include_file] [-p pad_value] [-Q precision]\n"
-	    "              [-r depth] [-W warning] [-X max_errors] <file>\n"
+	    "              [-r depth] [-s state] [-W warning] [-X max_errors] <file>\n"
```
Comment (ISSOtm) [inline src/asm/main.cpp:94]: How about `[-s features:state_file]`?

**Review by ISSOtm — DISMISSED** [review]

File: include/asm/output.hpp (hunk 2 above)
Comment (Rangi42) [reply include/asm/output.hpp:34]: Because `out_WriteState` may reassign it, `name = "<stdout>"`.

File: man/rgbasm.1 [hunk 5]
```diff
@@ -143,8 +144,45 @@ The argument may start with a
...
+.Ql Ic def Ar name Ic = Ar value .
+.It Cm equs
+Write all string constants as
+.Ql Ic def Ar name Ic equs Qq Ar value .
+.It Cm char
+Write all characters as
+.Ql Ic charmap Ar name , Ar value .
+.It Cm macro
+Write all macros as
+.Ql Ic macro Ar name ... Ic endm .
```
Comment (ISSOtm) [inline man/rgbasm.1:177]:
```suggestion
.Ql Ic macro Ar name No ... Ic endm .
```

File: src/asm/charmap.cpp [hunk 6]
```diff
@@ -17,59 +18,73 @@
...
+	size_t next[256]; // Indexes of where to go next, 0 = nowhere
 
 	bool isTerminal() const { return !value.empty(); }
 };
 
 struct Charmap {
 	std::string name;
-	std::vector<CharmapNode> nodes; // first node is reserved for the root node
+	std::vector<CharmapNode> nodes;                   // first node is reserved for the root node
+	std::unordered_map<size_t, std::string> mappings; // keys are indexes of terminal nodes
```
Comment (ISSOtm) [inline src/asm/charmap.cpp:31]:
```suggestion
	std::unordered_map<size_t, std::string> mappings; // keys are indexes of terminal nodes (FIXME: strictly speaking, this is redundant, we could walk the trie instead)
```

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Very good! This is almost ready for prime time.

**Review by ISSOtm — APPROVED** [review]
Perfect! Thank you *very* much :)

**Pull request merged**